src/byteset/mod.rs

```diff
@@ -0,0 +1,142 @@
+                        .is_some()
+                })
+            }
+        }
+    }
+}
+
+#[inline]
+pub(crate) fn find_not(haystack: &[u8], byteset: &[u8]) -> Option<usize> {
+    if haystack.is_empty() {
+        return None;
+    }
+    match byteset.len() {
+        0 => return Some(0),
+        1 => scalar::inv_memchr(byteset[0], haystack),
+        2 => scalar::forward_search_bytes(haystack, |b| {
+            b != byteset[0] && b != byteset[1]
+        }),
+        3 => scalar::forward_search_bytes(haystack, |b| {
+            b != byteset[0] && b != byteset[1] && b != byteset[2]
+        }),
+        n => {
+            if should_use_table(n) {
+                let table = build_table(byteset);
+                scalar::forward_search_bytes(haystack, |b| {
+                    table[b as usize] == 0
+                })
+            } else {
+                scalar::forward_search_bytes(haystack, |b| {
+                    scalar::forward_search_bytes(byteset, |sb| b != sb)
+                        .is_some()
+                })
+            }
+        }
+    }
+}
+#[inline]
+pub(crate) fn rfind_not(haystack: &[u8], byteset: &[u8]) -> Option<usize> {
+    if haystack.is_empty() {
+        return None;
+    }
+    match byteset.len() {
+        0 => return Some(haystack.len() - 1),
+        1 => scalar::inv_memrchr(byteset[0], haystack),
+        2 => scalar::reverse_search_bytes(haystack, |b| {
+            b != byteset[0] && b != byteset[1]
+        }),
+        3 => scalar::reverse_search_bytes(haystack, |b| {
+            b != byteset[0] && b != byteset[1] && b != byteset[2]
+        }),
+        n => {
+            if should_use_table(n) {
+                let table = build_table(byteset);
+                scalar::reverse_search_bytes(haystack, |b| {
+                    table[b as usize] == 0
+                })
+            } else {
+                scalar::reverse_search_bytes(haystack, |b| {
+                    scalar::forward_search_bytes(byteset, |sb| b != sb)
+                        .is_some()
+                })
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    quickcheck! {
+        fn qc_byteset_forward_matches_naive(haystack: Vec<u8>, needles: Vec<u8>) -> bool {
+            super::find(&haystack, &needles) == haystack.iter().position(|b| needles.contains(b))
+        }
+        fn qc_byteset_backwards_matches_naive(haystack: Vec<u8>, needles: Vec<u8>) -> bool {
+            super::rfind(&haystack, &needles) == haystack.iter().rposition(|b| needles.contains(b))
+        }
+        fn qc_byteset_forward_not_matches_naive(haystack: Vec<u8>, needles: Vec<u8>) -> bool {
+            super::find_not(&haystack, &needles) == haystack.iter().position(|b| !needles.contains(b))
+        }
+        fn qc_byteset_backwards_not_matches_naive(haystack: Vec<u8>, needles: Vec<u8>) -> bool {
+            super::rfind_not(&haystack, &needles) == haystack.iter().rposition(|b| !needles.contains(b))
+        }
+    }
```

Comment:
I kind of feel like this might not be enough test coverage. Most of the code here is simple enough that I'd probably be okay with it, but the `inv_memchr1` function is not as simple and probably deserves more testing. A particularly important part of testing code like `inv_memchr1` is to make sure all offsets are tested, which is what the memchr crate does: https://github.com/BurntSushi/rust-memchr/blob/665fffcc9506ac1cd93339470d2ca4b018e648c6/src/tests/mod.rs#L353-L374

I realize this is perhaps more work than you wanted to take on. I would totally be okay with leaving the extra tests out of this PR, but I'd probably insist on dropping `inv_memchr1` in addition to that. (You could just throw it up into another PR without tests, and I'd be happy to add tests to it later when I get a chance. It's up to you and your time/desire budget. :-))

Reply:
Adding more tests doesn't bother me, I was just being lazy.

Reply:
I've added what I think amounts to equivalent tests to what you have for memchr 1 byte, although it's possible it's still not ideal, let me know.